flink-runtime/src/main/java/org/apache/flink/runtime/operators/JoinDriver.java

```diff
@@ -128,17 +130,22 @@ public void prepare() throws Exception{
 				ConfigConstants.DEFAULT_RUNTIME_HASH_JOIN_BLOOM_FILTERS);
 
 		// create and return joining iterator according to provided local strategy.
-		if (objectReuseEnabled) {
-			switch (ls) {
-				case INNER_MERGE:
-					this.joinIterator = new ReusingMergeInnerJoinIterator<>(in1, in2, 
+		if (reset) {
+			resetForIterativeTasks(in1, in2, serializer1, serializer2, comparator1, comparator2, pairComparatorFactory);
+			reset = false;
+		}
+		if (joinIterator == null) {
```

Comment:
I am not very sure how to make use of this Resettable framework without having to use a boolean reset or checking for null for the joinIterator. Ideally the reset() API is not needed.
In a normal way this is the flow

```
initialize() (for 1st iteration)
  or
reset() (for all other iterations)
followed by
setup()
prepare()
run()
cleanup().

```
Now if i try to use `reset()` to reset my iterator using the current taskContext in1 and in2. Then when the `run() `is called for the next iteration we don't have the latest in1 and in2. As said above see test case `PageRankITCase#testPageRankSmallNumberOfIterations`. This test fails. It expects the in1 and in2 to be updated after all the `resetInput()` happens from the batch task.

In that case the best way is to set a boolean in `reset()` and in prepare you do the actual iterator reset and reset the boolean flag after that. (here allow all the `prepare()` to do the actual initializataion).

In case I move all the current code in `prepare() `to `initialize()` then am not sure how to use the `prepare() and reset()` method to ensure that we actually do the iterator reset. Here again I need a reset boolean flag to do that.